pkg/daemon/certificate_writer.go

```diff
@@ -101,92 +117,230 @@ func (dn *Daemon) syncControllerConfigHandler(key string) error {
-				}
-				// we need to make sure we honor the mode of that file
-				if err := writeFileAtomically(bundle, data, mode, Finfo.Mode(), -1, -1); err != nil {
-					return err
-				}
+		if controllerConfig.Annotations[ctrlcommon.ServiceCARotateAnnotation] == ctrlcommon.ServiceCARotateTrue && dn.node.Annotations[constants.ControllerConfigSyncServerCA] != controllerConfig.Annotations[ctrlcommon.ServiceCARotateAnnotation] {
+			cm, cmErr = dn.kubeClient.CoreV1().ConfigMaps("openshift-machine-config-operator").Get(context.TODO(), "kubeconfig-data", v1.GetOptions{})
+			if cmErr != nil {
+				klog.Errorf("Error retrieving kubeconfig-data. %v", cmErr)
 			} else {
-				if err := writeFileAtomicallyWithDefaults(bundle, data); err != nil {
-					return err
+				data, err = cmToData(cm, "ca-bundle.crt")
+				if err != nil {
+					klog.Errorf("kubeconfig-data ConfigMap not populated yet. %v", err)
+				} else if data != nil {
+					currentKC := clientcmdv1.Config{}
+					kcBytes, err := os.ReadFile(kubeConfigPath)
+					if err != nil {
+						return err
+					}
+					if kcBytes != nil {
+						klog.Info("We should be writing this file")
+						err = yaml.Unmarshal(kcBytes, &currentKC)
+						if err != nil {
+							return fmt.Errorf("Could not unmarshal kubeconfig into struct. Data: %s, Error: %v", string(kcBytes), err)
+						}
+						klog.Infof("Here is KC data: %b", kcBytes)
+						var newData []byte
+						if currentKC.Clusters == nil {
+							return errors.New("Clusters cannot be nil")
+						}
+						klog.Infof("Here is Clusters[0] %v", currentKC.Clusters[0])
+						currentKC.Clusters[0].Cluster.CertificateAuthorityData = data
+						// take data from currentKC which now has new ca in it and marshal it
+						newData, err = json.Marshal(currentKC)
+						if err != nil {
+							return fmt.Errorf("Could not marshal kubeconfig into bytes. Error: %v", err)
+						}
+						// take old kubeconfig-data cm and marshal it
+						oldCMBin, err := json.Marshal(cm)
+						if err != nil {
+							return fmt.Errorf("Could not marshal old configmap %v", err)
+						}
+						// newCM == old kubeconfig-data and add in a kubeconfig field
+						newCM := cm
+						newCM.BinaryData["kubeconfig"], err = json.Marshal(currentKC)
+						if err != nil {
+							return fmt.Errorf("Could not marshal new kubeconfig %v", err)
+						}
+						// marshal it
+						newCMBin, err := json.Marshal(newCM)
+						if err != nil {
+							return fmt.Errorf("Could not marshal new configmap %v", err)
+						}
+						// patch kc data with old kubeconfig-data, new kubeconfig-data, and old kubeconfig-data
+						patch, err := jsonmergepatch.CreateThreeWayJSONMergePatch(oldCMBin, newCMBin, oldCMBin)
+						if err != nil {
+							return fmt.Errorf("Could not create patch for  kubeconfig-data %v", err)
+						}
+						_, err = dn.kubeClient.CoreV1().ConfigMaps(ctrlcommon.MCONamespace).Patch(context.TODO(), "kubeconfig-data", types.MergePatchType, patch, v1.PatchOptions{})
```

Comment:
Why are we patching the kubeconfig-data back here? Shouldn't the daemon just read from the configmap?